Skulltula Token message enhancements: add kill count and auto-dismiss message

vaguerant opened the pull request:
**EDIT:** PR has changed substantially since this original comment; the code explanations are outdated and no longer relevant. Now, the PR adds two features:
- if the player has Dog's new Skulltula kill count toggle (`gInjectSkulltulaCount`) enabled, the message is replaced with a new message reporting how many tokens the player has collected, ala the one from OoT Redux (supports all three languages that are currently in SoH)
- if the player has Argorath's Skulltula message freeze toggle (`gSkulltulaFreeze`) enabled, the message now displays like all the other non-freeze messages (e.g. the ones when an egg hatches, when Link is underwater without the Zora Tunic or in a hot area without the Goron Tunic), with instant text and dismissing itself automatically after 3 seconds

Both features interact cleanly: you can have injected kill counts but not the freeze toggle, the freeze toggle but not injected kill counts, both or neither. All credit to Dog for these changes.

Original PR filing text continues below. **/EDIT**

Addendum to @GreatArgorath's #365. Sorry about yet another addendum PR, I was really just fooling around to see if this was achievable and it turned out it was. I checked in with Argorath before PRing, although they haven't seen the code, so don't take that as approval of what's actually *in* this PR. 😄

Quick explanation of what it's doing:
```
            char* goldSkullPatch = (char*)malloc(sizeof(char) * (file->messages[i].msg.size() + 3));
            file->messages[i].msg.copy(goldSkullPatch, file->messages[i].msg.size(), 0);
```
Creating the new automated version of the message by copying the original. That `+ 3` on the end is making the message three bytes longer than the original Skulltula Token message, because of the extra control code bytes this patch adds at the start (1) and end (2).

```c
            memmove(goldSkullPatch+1, goldSkullPatch, file->messages[i].msg.size());
            goldSkullPatch[00] = 0x08;
```
This moves the message over one byte in memory, because the "instant" control code `08` has to come at the start of the message. This is probably the part I'm most wary about; I don't know if there's a better way to handle moving the whole message over 1 byte and prefixing it with the control code. If this is bad, I'm open to other ideas. Leaving this change out entirely means the message slowly fills in like normal text boxes do, where all the other automated text boxes like this (egg hatched, you can't breathe underwater, it's too hot in here, etc.) are instant.

```c
                goldSkullPatch[79] = 0x0E;
                goldSkullPatch[80] = 60;
                goldSkullPatch[81] = 0x02;
```
This (and the German/French equivalents) are replacing the original final message byte `02` with `0E 3C` (dismiss message after 60 frames) **then** `02`. I chose 60 or three seconds at 20 FPS as that's what the base game uses for its most common instant messages (too hot/underwater). Others like the egg hatching (99 frames/4.95 seconds) do use longer times, but for a message you see 100 times in the course of a complete run, you probably don't need that long to understand what the message is telling you.

The rest is pretty self-explanatory.

**EDIT:** Forgot to credit @earthcrafterman's #354, which, in addition to their advice on Discord. forms the vast bulk of this PR as it provided the example for patching a message from the OTR.

vaguerant committed:
Make Skulltula Token message automated

GreatArgorath commented:
@Dog Your skulltula counter addition might be a good fit for this, it'd make it identical to the Redux feature then!

vaguerant committed:
Don't need + 3 here

jbodner09 commented:
Would it be helpful to rename this PR to "Make Skulltula Token message display instantly and dismiss automatically"?  "Automated" is a little unclear to me.

vaguerant commented:
@Dog's closed PR #468 for reporting the current token count in the Skulltula message uses an entirely different method to replace the message which looks much easier for making more substantial changes, rather than patching one byte at a time as I do here. Dog suggested merging their behavior into this PR, but since it's essentially unrelated to how this one operates, I suggested they could just submit their own PR incorporating these control code changes into their solution if they prefer.

For now, I'm converting this to a draft until the way forward is chosen.

Dog committed:
Feature: Count Golden Skulltulas in item pickup message

Co-authored-by: vaguerant <vaguerant@users.noreply.github.com>

Dog committed:
Fix logic

Dog committed:
Set textBoxType since Message_FindMessage is not used

Dog committed:
SelectedLanguage suffers from an initialization order issue, it will not give the correct result until cvars map is fully initialized.

If called before fully allocated (which is what can happen in the current code state), it will throw an exception.

Dog committed:
Handle combinations of freeze, count skulltula, and vanilla messages.

vaguerant committed:
cleanups, whitespace

Dog committed:
Update German text

vaguerant committed:
rm stray line break, definitely not prodding jenkins

Dog committed:
Merge branch 'develop' into tokenpatch

Dog committed:
Update translations, remove 0xB5 case

Dog committed:
Add back in newline at end of OTRGlobals.h

louist103 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1290,3 +1290,57 @@ extern "C" int Controller_ShouldRumble(size_t i) {
+    std::string postfix(freeze ? "\x0E\x3C\x02" : "\x02");
+    std::string str;
+
+    if (src == NULL) {
```
```suggestion
    if (src == nullptr) {
```

Dog replied:
Not sure why but github does not seem to notice this commit suggestion was applied. It is still listed as blocking for merge

louist103 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1290,3 +1290,57 @@ extern "C" int Controller_ShouldRumble(size_t i) {
     return 0;
 }
+
+extern "C" int GetSkulltulaCountMessage(char* buffer, char* src, const int maxBufferSize) {
```
Why the return? It looks like it always returns zero and its never used anyway.

Dog replied:
This was a mistake and the return was supposed to be used by the caller of this function. Instead a strlen was added from some debugging that was done and was not replaced.

Dog committed:
Update soh/soh/OTRGlobals.cpp

Co-authored-by: louist103 <35883445+louist103@users.noreply.github.com>

Dog committed:
Update z_message_PAL.c

Properly use the value of the GetSkulltulaCountMessage function.

vaguerant committed:
Merge branch 'develop' into tokenpatch

vaguerant committed:
Merge branch 'develop' into tokenpatch

vaguerant committed:
Merge branch 'develop' into tokenpatch

vaguerant committed:
Merge branch 'develop' into tokenpatch

vaguerant committed:
Merge fix

leggettc18 commented:
Most everything from this merge request is already in develop by this point, the only thing that is missing is that the text box disappears automatically in this PR, In develop you still have to hit the button. Depending on how #767 turns out the text from the skulltula messages may be relocating anyway, and that would be a good time to implement the auto-close I think. That PR is still evolving though so it may turn out that Skulltula text doesn't end up involved, it's a bit too early to tell.

vaguerant commented:
Closing this as some of it is already in dev via rando, and #831 covers most of the rest, barring freeze behavior.
